tests/unit/test_data_retriever.py

```diff
@@ -0,0 +1,374 @@
+    mocked_load_tsv.assert_called_once_with(
+        api._study.layout.fpath_harmonized, index_cols=api._index_cols_phenotypes
+    )
+    mocked_filter_with_manifest.assert_called_once_with(df_harmonized)
+
+    assert list(df_phenotypes.columns) == phenotypes
+
+
+def test_get_derivatives(api: NipoppyDataRetriever, mocker: pytest_mock.MockFixture):
+    mocked_check_derivatives_arg = mocker.patch(
+        "nipoppy._data_retriever._check_derivatives_arg"
+    )
+    mocked_get_derivatives_table = mocker.patch.object(
+        api,
+        "_get_derivatives_table",
+        # return string instead of df for testing purposes
+        side_effect=(lambda name, version, pattern: f"df_{pattern}"),
+    )
+    mocked_pd_concat = mocker.patch("pandas.concat", return_value=pd.DataFrame())
+    mocked_filter_with_manifest = mocker.patch.object(
+        api, "_filter_with_manifest", return_value=pd.DataFrame()
+    )
+
+    derivatives = [
+        ("pipeline1", "v1.0", "pattern1"),
+        ("pipeline2", "v2.0", "pattern2"),
+    ]
+    df_derivatives = api.get_derivatives(derivatives=derivatives)
+
+    mocked_check_derivatives_arg.assert_called_once_with(derivatives)
+    assert mocked_get_derivatives_table.call_count == len(derivatives)
+    for derivative_spec in derivatives:
+        mocked_get_derivatives_table.assert_any_call(*derivative_spec)
+    mocked_pd_concat.assert_called_once_with(
+        ["df_pattern1", "df_pattern2"], axis="columns", join="outer"
+    )
+    assert id(mocked_filter_with_manifest.call_args[0][0]) == id(
+        mocked_pd_concat.return_value
+    )
+    assert id(df_derivatives) == id(mocked_filter_with_manifest.return_value)
+
+
+def test_get_tabular_data(api: NipoppyDataRetriever, mocker: pytest_mock.MockFixture):
+    mocked_get_phenotypes_table = mocker.patch.object(
+        api, "get_phenotypes", return_value="df_phenotypes"
+    )
+    mocked_get_derivatives_table = mocker.patch.object(
+        api, "get_derivatives", return_value="df_derivatives"
+    )
+    mocked_pd_concat = mocker.patch("pandas.concat", return_value=pd.DataFrame())
+
+    phenotypes = ["nb:Age"]
+    derivatives = [
+        ("pipeline1", "v1.0", "pattern1"),
+        ("pipeline2", "v2.0", "pattern2"),
+    ]
+    df_tabular = api.get_tabular_data(phenotypes=phenotypes, derivatives=derivatives)
+
+    mocked_get_phenotypes_table.assert_called_once_with(phenotypes)
+    mocked_get_derivatives_table.assert_called_once_with(derivatives)
+    mocked_pd_concat.assert_called_once_with(
+        ["df_phenotypes", "df_derivatives"], axis="columns", join="outer"
+    )
+    assert id(df_tabular) == id(mocked_pd_concat.return_value)
+
+
+@pytest.mark.parametrize(
+    "phenotypes,derivatives,get_phenotypes_called,get_derivatives_called",
+    [
+        (["nb:Age"], None, True, False),
+        (
+            None,
+            [("pipeline1", "v1.0", "pattern1")],
+            False,
+            True,
+        ),
```

Comment:
I would prefer the params to have the same formatting (for readability), either both one-liners or mutli-lines